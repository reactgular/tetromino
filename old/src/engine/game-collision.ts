import {GameBuffer} from './game-buffer';
import {GameTransform} from './game-transform';

/**
 * Checks if a coordinate is inside the buffer.
 */
export function gameCollideBuffer(
    {width, height}: GameBuffer,
    x: number,
    y: number
): boolean {
    return x < 0 || x >= width || y < 0 || y >= height;
}

/**
 * Checks if a coordinate hits a block in the buffer.
 */
export function gameCollideBlock(
    {rows}: GameBuffer,
    x: number,
    y: number
): boolean {
    return rows[y].cells[x].type !== undefined;
}

/**
 * Checks if the coordinates of a piece cause any collision.
 */
export function gameCollision(
    {x, y, piece}: GameTransform,
    buffer: GameBuffer
): boolean {
    for (let i = 0; i < piece.length; i++) {
        const [offset_x, offset_y] = piece[i];
        const trans_x = x + offset_x;
        const trans_y = y + offset_y;
        if (
            gameCollideBuffer(buffer, trans_x, trans_y) ||
            gameCollideBlock(buffer, trans_x, trans_y)
        ) {
            return true;
        }
    }
    return false;
}
