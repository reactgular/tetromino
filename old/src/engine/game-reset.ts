import {randTetros} from '../components/particles/utilities.types';
import {GAME_INITIAL_STATE, GameModel} from '../store/game/game-model';
import {gameBufferEmpty} from './game-buffer';

export const BUFFER_WIDTH = 10;

export const BUFFER_HEIGHT = 20;

export const gameReset = (next_max: number): GameModel => {
    const buffer = gameBufferEmpty(BUFFER_WIDTH, BUFFER_HEIGHT);
    return {
        ...GAME_INITIAL_STATE,
        next: randTetros(next_max),
        buffer: buffer,
        screen: Array(buffer.width * buffer.height).fill({}),
        level: 1,
        score: 0,
        player: undefined,
        ghost: undefined
    };
};
