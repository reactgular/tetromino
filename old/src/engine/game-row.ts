import {gameBufferCells, GameCell} from './game-cell';

/**
 * Defines a row across the screen.
 */
export interface GameRow {
    cells: Array<GameCell>;

    removed?: boolean;
}

/**
 * Generates an empty collection of rows.
 */
export function* gameBufferRows(
    width: number,
    height: number
): IterableIterator<GameRow> {
    for (let i = 0; i < height; i++) {
        yield {cells: Array.from(gameBufferCells(width))};
    }
}
