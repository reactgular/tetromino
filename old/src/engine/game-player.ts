import {GameBuffer, gameBufferRowSolid} from './game-buffer';
import {gameCollision} from './game-collision';
import {
    PIECE_I,
    PIECE_J,
    PIECE_L,
    PIECE_O,
    PIECE_S,
    PIECE_T,
    PIECE_Z
} from './game-pieces';
import {GameRow} from './game-row';
import {TetrominosType} from './game-tetrominos';
import {GameRotator, gameTransform, GameTranslator} from './game-transform';

/**
 * Defines a player piece with location.
 */
export interface GamePlayer {
    rotate: number;

    type: TetrominosType;

    x: number;

    y: number;
}

/**
 * Defines a ghost piece location.
 */
export type GameGhost = Pick<GamePlayer, 'x' | 'y'>;

/**
 * An array from the GameDesigner.
 */
export type GamePiece = Array<[number, number]>;

/**
 * A single game piece in different rotations.
 */
export type GamePieceRotations = Array<GamePiece>;

/**
 * Map of game pieces grouped by their type.
 */
export const GAME_PIECES: Record<string, GamePieceRotations> = {
    [TetrominosType.O]: PIECE_O,
    [TetrominosType.I]: PIECE_I,
    [TetrominosType.S]: PIECE_S,
    [TetrominosType.Z]: PIECE_Z,
    [TetrominosType.L]: PIECE_L,
    [TetrominosType.J]: PIECE_J,
    [TetrominosType.T]: PIECE_T
};

/**
 * Creates a player piece in the center of the buffer. Piece is positioned
 * as close as possible to the top of the buffer.
 */
export function gamePlayerCreate(
    type: TetrominosType,
    width: number
): GamePlayer {
    const piece = GAME_PIECES[type][0];
    const y = piece.some(([offset_x, offset_y]) => offset_y < 0) ? 1 : 0;
    return {
        x: Math.floor(width / 2),
        y,
        type,
        rotate: 0
    };
}

/**
 * Bakes the player's piece into the buffer.
 */
export function gamePlayerDrop(
    {x, y, type, rotate}: GamePlayer,
    buffer: GameBuffer
): 0 | 1 | 2 | 3 | 4 {
    const piece = GAME_PIECES[type][rotate];
    const rows = new Set<GameRow>();
    for (let i = 0; i < piece.length; i++) {
        const [offset_x, offset_y] = piece[i];
        const row = buffer.rows[y + offset_y];
        row.cells[x + offset_x] = {type};
        rows.add(row);
    }
    let lines = 0;
    rows.forEach((row) => {
        if (gameBufferRowSolid(row)) {
            row.removed = true;
            lines++;
        }
    });
    return lines as 0 | 1 | 2 | 3 | 4;
}

/**
 * Moves/rotates a player only if there is no collision.
 */
export function gamePlayerTransform(
    buffer: GameBuffer,
    player: GamePlayer,
    translator: GameTranslator,
    rotator: GameRotator
) {
    const trans = gameTransform(player, translator, rotator);
    if (!gameCollision(trans, buffer)) {
        player.rotate = trans.rotate;
        player.x = trans.x;
        player.y = trans.y;
    }
}
