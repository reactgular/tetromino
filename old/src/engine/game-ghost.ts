import {GameBuffer} from './game-buffer';
import {gameCollision} from './game-collision';
import {GameGhost, GamePlayer} from './game-player';
import {gameTransform, rotateNoop, translateDown} from './game-transform';

/**
 * Drops a piece to find the ghost location.
 */
export const gameGhost = (
    buffer: GameBuffer,
    {x, y, type, rotate}: GamePlayer
): GameGhost | undefined => {
    let ghost: GameGhost | undefined;
    while (true) {
        const trans = gameTransform(
            {x: ghost?.x || x, y: ghost?.y || y, type, rotate},
            translateDown,
            rotateNoop
        );
        if (gameCollision(trans, buffer)) {
            return ghost;
        } else {
            ghost = {x: trans.x, y: trans.y};
        }
    }
};
