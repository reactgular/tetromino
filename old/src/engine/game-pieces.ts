import {GamePieceRotations} from './game-player';

export const PIECE_O: GamePieceRotations = [
    [
        [-1, 0],
        [0, 0],
        [-1, 1],
        [0, 1]
    ]
];

export const PIECE_I: GamePieceRotations = [
    [
        [-2, 0],
        [-1, 0],
        [0, 0],
        [1, 0]
    ],
    [
        [0, -1],
        [0, 0],
        [0, 1],
        [0, 2]
    ]
];

export const PIECE_S: GamePieceRotations = [
    [
        [0, -1],
        [0, 0],
        [1, 0],
        [1, 1]
    ],
    [
        [0, 0],
        [1, 0],
        [-1, 1],
        [0, 1]
    ]
];

export const PIECE_Z: GamePieceRotations = [
    [
        [1, -1],
        [0, 0],
        [1, 0],
        [0, 1]
    ],
    [
        [-1, 0],
        [0, 0],
        [0, 1],
        [1, 1]
    ]
];

export const PIECE_L: GamePieceRotations = [
    [
        [1, -1],
        [-1, 0],
        [0, 0],
        [1, 0]
    ],
    [
        [-1, -1],
        [0, -1],
        [0, 0],
        [0, 1]
    ],
    [
        [-1, 0],
        [0, 0],
        [1, 0],
        [-1, 1]
    ],
    [
        [0, -1],
        [0, 0],
        [0, 1],
        [1, 1]
    ]
];

export const PIECE_J: GamePieceRotations = [
    [
        [-1, -1],
        [-1, 0],
        [0, 0],
        [1, 0]
    ],
    [
        [0, -1],
        [0, 0],
        [-1, 1],
        [0, 1]
    ],
    [
        [-1, 0],
        [0, 0],
        [1, 0],
        [1, 1]
    ],
    [
        [0, -1],
        [1, -1],
        [0, 0],
        [0, 1]
    ]
];

export const PIECE_T: GamePieceRotations = [
    [
        [0, -1],
        [-1, 0],
        [0, 0],
        [1, 0]
    ],
    [
        [0, -1],
        [-1, 0],
        [0, 0],
        [0, 1]
    ],
    [
        [-1, 0],
        [0, 0],
        [1, 0],
        [0, 1]
    ],
    [
        [0, -1],
        [0, 0],
        [1, 0],
        [0, 1]
    ]
];
