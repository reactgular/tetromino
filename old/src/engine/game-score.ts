import {SOUND_LEVEL, SOUND_SCORE} from '../components/particles/audio.types';
import {GameModel} from '../store/game/game-model';
import {GAME_PIECES, GamePiece} from './game-player';
import {TetrominosType} from './game-tetrominos';

const LEVEL_1_PER_LINE = {
    1: (n: number) => 40,
    2: (n: number) => 100,
    3: (n: number) => 300,
    4: (n: number) => 1200
};

const LEVEL_2_PER_LINE = {
    1: (n: number) => 80,
    2: (n: number) => 200,
    3: (n: number) => 600,
    4: (n: number) => 2400
};

const LEVEL_3_PER_LINE = {
    1: (n: number) => 120,
    2: (n: number) => 300,
    3: (n: number) => 900,
    4: (n: number) => 3600
};

const LEVEL_N_PER_LINE = {
    1: (n: number) => 40 * (n + 1),
    2: (n: number) => 100 * (n + 1),
    3: (n: number) => 300 * (n + 1),
    4: (n: number) => 1200 * (n + 1)
};

/**
 * Calculates the score for lines cleared.
 */
export const gamePointsLines = (
    level: number,
    lines: 0 | 1 | 2 | 3 | 4
): number => {
    if (lines !== 0) {
        if (level === 1) {
            return LEVEL_1_PER_LINE[lines](level);
        } else if (level === 2) {
            return LEVEL_2_PER_LINE[lines](level);
        } else if (level >= 3 && level < 10) {
            return LEVEL_3_PER_LINE[lines](level);
        } else if (level >= 10) {
            return LEVEL_N_PER_LINE[lines](level);
        }
    }
    return 0;
};

/**
 * Calculates the score for soft dropping a piece.
 */
export const gamePointsPiece = (piece: GamePiece): number => {
    return piece.length;
};

/**
 * Calculates the score for landing a piece.
 */
export const gamePoints = (
    level: number,
    type: TetrominosType,
    lines: 0 | 1 | 2 | 3 | 4
): number => {
    const lineScore = gamePointsLines(level, lines);
    const pieceScore = gamePointsPiece(GAME_PIECES[type][0]);
    return lineScore + pieceScore;
};

export type GameScore = Pick<
    GameModel,
    | 'score'
    | 'level'
    | 'lines_level'
    | 'lines_level_up'
    | 'lines_total'
    | 'player_freeze'
>;

/**
 * Computes the score and level changes for a collision.
 */
export const gameScore = (
    state: GameScore,
    type: TetrominosType,
    lines: 0 | 1 | 2 | 3 | 4
): string | undefined => {
    state.score = state.score + gamePoints(state.level, type, lines!);
    state.lines_level = state.lines_level + lines!;
    if (state.lines_level >= state.lines_level_up) {
        state.lines_total += state.lines_level;
        state.lines_level = 0;
        state.level++;
        return SOUND_LEVEL;
    }
    if (!state.player_freeze) {
        return SOUND_SCORE;
    }
};
