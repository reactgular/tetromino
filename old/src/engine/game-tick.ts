import {
    GameBuffer,
    gameBufferFindRemoved,
    gameBufferRemoveRow
} from './game-buffer';
import {gameCollision} from './game-collision';
import {GamePlayer, gamePlayerDrop} from './game-player';
import {gameTransform, rotateNoop, translateDown} from './game-transform';

/**
 * Results of moving the player's piece.
 */
export interface GameTickResult {
    /**
     * Was the piece dropped into the buffer?
     */
    collision: boolean;

    /**
     * (when collision is true) number of lines flagged for removal.
     */
    lines?: 0 | 1 | 2 | 3 | 4;
}

/**
 * Performs a single frame of animation of the player piece moving.
 */
export function gameTickPlayer(
    buffer: GameBuffer,
    player: GamePlayer
): GameTickResult {
    const trans = gameTransform(player, translateDown, rotateNoop);
    if (gameCollision(trans, buffer)) {
        const lines = gamePlayerDrop(player, buffer);
        return {collision: true, lines};
    }
    player.y = trans.y;
    return {collision: false};
}

/**
 * Drops the play down all the way until collision.
 */
export function gameDropPlayer(buffer: GameBuffer, player: GamePlayer) {
    while (true) {
        const trans = gameTransform(player, translateDown, rotateNoop);
        if (gameCollision(trans, buffer)) {
            break;
        } else {
            player.y = trans.y;
        }
    }
}

/**
 * Performs a single frame of animation of the rows collapsing down by gravity.
 */
export function gameTickRows(buffer: GameBuffer): boolean {
    const removeIndex = gameBufferFindRemoved(buffer);
    if (removeIndex === -1) {
        return false;
    }
    gameBufferRemoveRow(buffer, removeIndex);
    return true;
}
