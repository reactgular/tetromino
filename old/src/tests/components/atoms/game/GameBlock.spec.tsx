import {render} from '@testing-library/react';
import renderer from 'react-test-renderer';
import {
    GAME_BLOCK_COLOR,
    GameBlock
} from '../../../../components/atoms/game/GameBlock';
import {TETROMINOS_TYPES} from '../../../../engine/game-tetrominos';

describe(GameBlock.name, () => {
    it('should render the same as defaults if type is empty', () => {
        const tree = renderer.create(<GameBlock />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should set the block color', () => {
        TETROMINOS_TYPES.map((type) => {
            const {container} = render(<GameBlock type={type} />);
            const $ = container.querySelector.bind(container);
            expect($('.block-color')).toBeTruthy();
            expect($(`.${GAME_BLOCK_COLOR[type]}`)).toBeTruthy();
        });
    });

    it('should not glow unless there is a type', () => {
        const {container} = render(<GameBlock glow={true} />);
        const $ = container.querySelector.bind(container);
        expect($('.color-glow')).toBeFalsy();
    });

    it('should glow the block without a color', () => {
        TETROMINOS_TYPES.map((type) => {
            const {container} = render(<GameBlock glow={true} type={type} />);
            const $ = container.querySelector.bind(container);
            expect($('.block-color')).toBeTruthy();
            expect($('.color-glow')).toBeTruthy();
            expect($(`.${GAME_BLOCK_COLOR[type]}`)).toBeFalsy();
        });
    });
});
