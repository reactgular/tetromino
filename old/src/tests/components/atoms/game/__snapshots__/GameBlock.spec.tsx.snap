// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBlock should render the same as defaults if type is empty 1`] = `
<div
  className="game-block"
  data-testid="game-block"
>
  <div
    className="block-color relative w-full h-full rounded-sm"
  />
</div>
`;
